hal/src/nRF52840/i2c_hal.cpp

```diff
@@ -310,28 +313,43 @@ int hal_i2c_init(hal_i2c_interface_t i2c, const hal_i2c_config_t* config) {
     os_thread_scheduling(false, nullptr);
     if (i2cMap[i2c].mutex == nullptr) {
         os_mutex_recursive_create(&i2cMap[i2c].mutex);
-    } else {
-        // Already initialized
-        os_thread_scheduling(true, nullptr);
-        return SYSTEM_ERROR_NONE;
-    }
+    } 
 
     // Capture the mutex and re-enable threading
     I2cLock lk(i2c);
     os_thread_scheduling(true, nullptr);
 
-    // Initialize internal data structure
+    if (i2cMap[i2c].configured) {
```

Comment:
I think one additional thing here so that we somewhat don't leak memory is either to free the passed buffers in config if `HAL_I2C_CONFIG_FLAG_FREEABLE` here on errors or at least specifically on `SYSTEM_ERROR_NOT_ENOUGH_DATA`, or do this in `TwoWire` perhaps.